### DISCUSSION TITLE
Reasons behind reportInvalidTypeVarUse

### DISCUSSION THREAD

**Discussion by ethframe** [microsoft/pyright#2111] [Q&A] [answered]
I have the following code:

```python
from typing import Generic, Optional, TypeVar

T = TypeVar('T')
V = TypeVar('V')


class A(Generic[T]):
    def __init__(self, a: int, b: Optional[T]):
        self.a = a
        self.b = b

    @classmethod
    def from_a(cls, other: "A[V]") -> "A[T]":
        return A(other.a, None)
```

It's intended use is as follows:

```python
a = A(1, 'a')  # Type of "a" is "A[str]"
b = A[int].from_a(a)  # Type of "b" is "A[int]"
```

However, pyright complains about a type variable `V` in a classmethod:

```
TypeVar "V" appears only once in generic function signature
```

I don't understand why there is such a restriction on the use of type variables.

**Comment by jakebailey** [comment, marked as answer]
The `V` is only in one place, so it doesn't actually do anything. Your code could be written as:

```python
from typing import Generic, Optional, TypeVar

T = TypeVar('T')

class A(Generic[T]):
    def __init__(self, a: int, b: Optional[T]):
        self.a = a
        self.b = b

    @classmethod
    def from_a(cls, other: "A[Any]") -> "A[T]":
        return A(other.a, None)
```

Because `V` was never used to actually constrain any other parts of `from_a`.

**Reply by ethframe** [reply]
I'm still not sure if this restriction is reasonable. What if we have code like this:

```python
from typing import Generic, TypeVar

T = TypeVar('T')
V = TypeVar('V')


class A(Generic[T, V]):
    pass


def a(v: A[T, T]) -> None:
    b(v)


def b(v: A[T, V]) -> None:
    a(v)  # Argument of type "A[T@b, V@b]" cannot be assigned to parameter "v" of type "A[T@a, T@a]" in function "a"
```

The signature of `b` could be thought of as guarantee that it cannot pass its argument to `a`. However, pyright does not allow this signature, and if we replace the type variables with `Any`, this guarantee is gone.

**Reply by JelleZijlstra** [reply]
Instead of `A[Any]` you should perhaps write `A[object]`. In general, `Any` silences the type checker and `object` means that objects of all types are acceptable.

**Reply by erictraut** [reply]
@ethframe, many people misunderstand how to use TypeVars. We added these errors to help them avoid the most common misuses. Mypy provides the same checks and errors. If you don't like how the `reportInvalidTypeVarUse` diagnostic check works, you can disable it globally or on a per-file basis.

@JelleZijlstra, you would need to write `A[Any]` if the type param is invariant. `A[object]` is acceptable if it's covariant.

**Reply by ethframe** [reply]
@erictraut, I still don't understand why using a type variable once in a signature is considered as a misuse. Mypy succesfully typecheck such code, and even PEP 484 contains examples when type variable appears once in a signature.

**Reply by erictraut** [reply]
This is a debatable issue, and I can (and have) argued both sides. For additional context, see this thread in the typing sig: https://mail.python.org/archives/list/typing-sig@python.org/thread/NRFNHGXHXPGBR6FP3TIOZZ6VS4XJZX6K/#NRFNHGXHXPGBR6FP3TIOZZ6VS4XJZX6K. We laded on our current solution because of a steady stream of bug reports based on a misunderstanding of how TypeVars are supposed to be used.

**Reply by ethframe** [reply]
Well, I ran into this problem again, with signature like `def foo(a: A[T, V], b: A[T, U]) -> A[T, V]`. Guess I'd better turn this diagnostic off.

However, I found a workaround, that actually looks like a bug:

```python
from typing import Generic, TypeVar

T = TypeVar('T')


class A(Generic[T]):
    ...


def foo(x: T) -> None: ...  # TypeVar "T" appears only once in generic function signature
def bar(x: A[T]) -> None: ...  # TypeVar "T" appears only once in generic function signature


B = TypeVar('B', bound=object)


def baz(x: B) -> None: ...  # TypeVar "B" appears only once in generic function signature
def qux(x: A[B]) -> None: ...
```

That is, if the type variable's bound is explicitly set to `object`, pyright doesn't emit error if the variable is used as a parameter of some generic type. @erictraut, is this behavior expected?

**Reply by erictraut** [reply]
Yes, this is intended behavior. If the TypeVar is explicitly bound or constrained, it is allowed to be used only once in a signature.
